content/blog/2020-12-18-db-compiler/index.md

```diff
@@ -0,0 +1,293 @@
+
+### Code Execution
+
+Once the C++ file has been generated, we invoke clang on the file with optimizations turned on (-O3). We allow any C++ header defined in the codebase to be included. This enables us to reuse classes defined in the codebase such as the ones for file reading. This generates a dynamic library that we load into the database with dlopen. We retrieve the aforementioned compute function via dlsym and then execute it.
+
+## Performance Evaluation
+
+### Benchmark Suite
+
+The specific benchmark suite I'm using is the TPC-H benchmark. The schema can be found in my code base in `tpch/schema.sql`. The underlying data is generated by an automated [tool](https://github.com/electrum/tpch-dbgen). This tool has various scale factors to run the queries at. Scale factor 1 means there is a total of 1 GB of data, 10 means 10 GB of data and so on. I'm running these experiments on my rather underpowered (and out of disk space) laptop so I ran it at scale factor 1.
+
+The queries involved are a mix of filters, group/by aggregations and joins. I've selected a set of 5 queries out of the whole benchmark to run the compiler on (queries 1, 3, 6, 11 and 12). These queries are unchanged except for the removal of subqueries, order by and having clauses as I don't support them. The remaining queries were mostly the same with those removed so I chose to omit them since I had to manually convert them to a physical query plan.
+
+### Baseline
+
+Postgres is a popular database that uses an interpretation-style execution engine so I thought it was fit to compare against as a baseline. Since my code is single-threaded, I set `max_parallel_workers` to 1. I also set `jit` off. Postgres includes a very basic compiler that will generate LLVM code for expressions only and not operators. However, since I am interested in comparing against the interpretation version, I set this off. I'm running Postgres 13.
+
+Postgres does store full table tuples packed on disk while I store them separately. This does mean we load less data which gives us an unfair advantage. To fix this, I created materialized views of each of the base tables with only the relevant loaded columns. For example, for Q1:
+```
+CREATE TABLE lineitem_01 AS (SELECT l_returnflag, l_linestatus, l_quantity, l_extendedprice, l_discount, l_tax, l_shipdate FROM lineitem);
+```
+Then, you can rewrite the query to use this newly created `lineitem_01` table. This way, Postgres and our system read the same amount of data from disk.
+
+I also verify correctness by checking that the query results are identical in Postgres.
+
+### Setup
+
+I'm running this on a Macbook Pro, Late 2013 with an i7-4750 HQ and 8GB of RAM. This is running Ubuntu 20.04 under the Windows Linux Subystem.
```

Comment:
Macbook -> MacBook
Windows Linux Subystem -> Windows Subsystem for Linux

It might be relevant to report whether you're using WSL or WSL2, since they are pretty different from each other (and pretty different from "real" Linux)…